### DISCUSSION TITLE
Components / higher-level primitives API

### DISCUSSION THREAD

**Discussion by atomiks** [atomiks/floating-ui#18] [Ideas]
As mentioned on the README and website, the plan of this project extends beyond just positioning. The goal is to let users craft accessible components easily, basically a modern version of Tippy.js.

These components will be written in vanilla JS which can be wrapped by any framework.

## Idea

- Export a generic `createFloating()` function, in which every component derives its behavior from. This is lower-level and not "accessible by default". This enables users to go beyond the components.
- Export higher-level components, like `createTooltip()`, `createPopover()`, etc. These are accessible by default.

## Architecture

How should these APIs look like? Configurable via options, or some other technique?

Should we make the components super opinionated? How should the generic `createFloating()` function work, should it be extensible via plugins?

**Comment by KubaJastrz** [comment]
> Export higher-level components, like createTooltip(), createPopover(), etc.

What kind of scope are you talking about when you mention this? There already exist some solutions for "headless" components, like https://headlessui.dev/ or Tippy. Both use popper for positioning, but I guess they are a little more complex.

I agree that devs need an easy and accessible solution to common elements, as the DOM doesn't offer this. But maybe some parts of this problem have already been solved.

**Reply by atomiks** [reply]
From what I can see, Headless UI doesn't use Popper, it does recommend to use it though. 

Tippy is outdated now with the way it bases itself around the CSS needing to be imported (styled). There's the headless option, but the whole thing could be rearchitected better imo. It's not tree-shakeable and modular enough. Also there's so much more it could do better, like handling nesting better, focus management, smarter cursor interactivity for nested submenus, etc. 

The goal here is to be lower-level than Tippy, ensuring it's powerful and extensible. I want it to be equally as functional/extensible as the core library I've made here, although it will likely need to be stateful and return an instance.
